### PR TITLE
make readme a bit more clear

### PR DESCRIPTION
**Pull request by jgao1025**
It is trying to solve #27, but I updated READ.md a bit more with my own opinion. Free feel to change, take some.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,11 +2,18 @@
 
 Before starting the installation process, you need:
 
-- An Ubuntu 20.04/22.04 machine with SSH credentials
-  - This must have ports 22 (SSH), 80 (HTTP), 443 (HTTPS) and 6443 (Kubernetes) exposed
+- Target building resources.
+  - Ubuntu 20.04/22.04 machine(s) with SSH credentials. 
+    - At least one, but also the script can also work for multiple nodes. The hostname of each node can be called `node0`, `node1`, etc.
+    - `node0` must have ports 22 (SSH), 80 (HTTP), 443 (HTTPS) and 6443 (Kubernetes) exposed. `node0` must have a static public IP address that allows domain name to point to. `node<x>` must have post 22 (SSH) exposed. All nodes are better to be in the same vlan so they can communicate with each other.
```
Comment (mrsimonemms) [inline README.md]: I'm not sure this is entirely correct. In my setup, there is a load balancer on all nodes HTTP and HTTPS. Also, if you use the `HA_CLUSTER`, then you can connect to the Kubernetes API on subsequent nodes.

I think for simplicity, just saying "expose these four ports" is ok.

File: README.md [hunk 2]
```diff
@@ -2,11 +2,18 @@
...
+    - At least 4 cores, 16GB RAM and 50GB storage.
```
Comment (mrsimonemms) [inline README.md]:
```suggestion
    - At least 4 cores, 16GB RAM and 100GB storage.
```

We had some people on Discord complaining below 100GB storage

File: README.md [hunk 3]
```diff
@@ -2,11 +2,18 @@
...
+  - A domain and some wildcard subdomains managed by cloudflare (free), gcp, or route53 [see price](https://aws.amazon.com/route53/pricing/). please see [here](https://www.gitpod.io/docs/configure/self-hosted/latest/installing-gitpod) about "DNS and TLS configured". These DNS services will have and manage free letsencrypt certificates for you. If you choose not using these commercial DNS services, you will need to use self-signed certificates and manage it by yourself.
```
Comment (mrsimonemms) [inline README.md]:
```suggestion
  - A domain and some wildcard subdomains managed by Cloudflare (free), GCP, or Route53 [see price](https://aws.amazon.com/route53/pricing/). Please see the "DNS and TLS configured" section in the [Gitpod docs](https://www.gitpod.io/docs/configure/self-hosted/latest/installing-gitpod) for more information. These DNS services will have and manage free Let's Encrypt certificates for you. If you choose not to use these commercial DNS services, you will need to use self-signed certificates and manage them manually.
```

File: README.md [hunk 4]
```diff
@@ -2,11 +2,18 @@
...
 - A `.env` file or environment variables with basic details about the environment.
   - We provide an example of such file [here](.env.example)
-- [Docker](https://docs.docker.com/engine/install) installed on your machine, or better, a [Gitpod workspace](https://gitpod.io/#https://github.com/MrSimonEmms/gitpod-k3s-guide) 😀
+- Building environment:
+  - build on a local linux machine - needs to install kubectl, Helm, K3sup, also check and clean up `~/.kube/` if there was config file before.
```
Comment (mrsimonemms) [inline README.md]:
```suggestion
  - Build on a local Linux machine - needs to install kubectl, Helm, K3sup. You may need to clean your `${HOME}/.kube` directory if there was a previous `gitpod-k3s` entry. 
```

File: README.md [hunk 5]
```diff
@@ -2,11 +2,18 @@
...
+    - At least 4 cores, 16GB RAM and 50GB storage.
+  - A domain and some wildcard subdomains managed by cloudflare (free), gcp, or route53 [see price](https://aws.amazon.com/route53/pricing/). please see [here](https://www.gitpod.io/docs/configure/self-hosted/latest/installing-gitpod) about "DNS and TLS configured". These DNS services will have and manage free letsencrypt certificates for you. If you choose not using these commercial DNS services, you will need to use self-signed certificates and manage it by yourself.
 - A `.env` file or environment variables with basic details about the environment.
   - We provide an example of such file [here](.env.example)
-- [Docker](https://docs.docker.com/engine/install) installed on your machine, or better, a [Gitpod workspace](https://gitpod.io/#https://github.com/MrSimonEmms/gitpod-k3s-guide) 😀
+- Building environment:
+  - build on a local linux machine - needs to install kubectl, Helm, K3sup, also check and clean up `~/.kube/` if there was config file before. 
+  - Or use [Docker](https://docs.docker.com/engine/install) installed on your machine and Docker file is at .gitpod/gitpod.Dockerfile
```
Comment (mrsimonemms) [inline README.md]:
```suggestion
  - Use [Docker](https://docs.docker.com/engine/install) installed on your machine and Docker file is at .gitpod/gitpod.Dockerfile
```

File: README.md [hunk 6]
```diff
@@ -2,11 +2,18 @@
...
+  - Or even better, use [Gitpod workspace](https://gitpod.io/#https://github.com/MrSimonEmms/gitpod-k3s-guide) 😀
```
Comment (mrsimonemms) [inline README.md]:
```suggestion
  - Even better, use [Gitpod workspace](https://gitpod.io/#https://github.com/MrSimonEmms/gitpod-k3s-guide) 😀
```

File: README.md [hunk 7]
```diff
@@ -2,11 +2,18 @@
...
+    - `node0` must have ports 22 (SSH), 80 (HTTP), 443 (HTTPS) and 6443 (Kubernetes) exposed. `node0` must have a static public IP address that allows domain name to point to. `node<x>` must have post 22 (SSH) exposed. All nodes are better to be in the same vlan so they can communicate with each other.
+    - At least 4 cores, 16GB RAM and 50GB storage.
+  - A domain and some wildcard subdomains managed by cloudflare (free), gcp, or route53 [see price](https://aws.amazon.com/route53/pricing/). please see [here](https://www.gitpod.io/docs/configure/self-hosted/latest/installing-gitpod) about "DNS and TLS configured". These DNS services will have and manage free letsencrypt certificates for you. If you choose not using these commercial DNS services, you will need to use self-signed certificates and manage it by yourself.
 - A `.env` file or environment variables with basic details about the environment.
   - We provide an example of such file [here](.env.example)
-- [Docker](https://docs.docker.com/engine/install) installed on your machine, or better, a [Gitpod workspace](https://gitpod.io/#https://github.com/MrSimonEmms/gitpod-k3s-guide) 😀
+- Building environment:
```
Comment (mrsimonemms) [inline README.md]:
```suggestion
- Building environment. You can either:
```

File: .env.example [hunk 8]
```diff
@@ -20,7 +20,8 @@ LETSENCRYPT_EMAIL=my@email
 MANAGED_DNS_PROVIDER=selfsigned
 
 # Cloudflare options
-CLOUDFLARE_API_KEY=""
+# Follow the instruction from "API Tokens" to create an API Token: https://cert-manager.io/docs/configuration/acme/dns01/cloudflare/
+CLOUDFLARE_API_TOKEN=""
```
Comment (mrsimonemms) [inline .env.example:24]: I'm never a fan of changing interfaces, but I think this is justified based on the original issue.

**Review by mrsimonemms — CHANGES_REQUESTED** [review]
A few stylistic changes, but I'm in favour of this change. Thank you for the contribution

File: README.md (hunk 1 above)
Comment (jgao1025) [reply README.md]: Yes, I agree it should be more general than to just talking about my own setting. I have applied the changes and also updated this part a bit.

**Comment by jgao1025** [comment]
> A few stylistic changes, but I'm in favour of this change. Thank you for the contribution

I don't know which part is blocked for the PR.  I am happy to fix anything that helps merging.

File: README.md [hunk 9]
```diff
@@ -2,11 +2,18 @@
...
 - A `.env` file or environment variables with basic details about the environment.
   - We provide an example of such file [here](.env.example)
-- [Docker](https://docs.docker.com/engine/install) installed on your machine, or better, a [Gitpod workspace](https://gitpod.io/#https://github.com/MrSimonEmms/gitpod-k3s-guide) 😀
+- Building environment. You can either:
+  - Build on a local Linux machine - needs to install kubectl, Helm, K3sup. You may need to clean your `${HOME}/.kube` directory if there was a previous `gitpod-k3s` entry. 
+  - Use [Docker](https://docs.docker.com/engine/install) installed on your machine and Docker file is at .gitpod/gitpod.Dockerfile.
+  - Even better, use [Gitpod workspace](https://gitpod.io/#https://github.com/MrSimonEmms/gitpod-k3s-guide)😀.
```
Comment (mrsimonemms) [inline README.md]:
```suggestion
  - Even better, use a [Gitpod workspace](https://gitpod.io/#https://github.com/MrSimonEmms/gitpod-k3s-guide)😀.
```

**Review by mrsimonemms — CHANGES_REQUESTED** [review]
@jgao1025 sorry, I'd missed that you'd done updates. Thanks for the prompt. One last thing on there.

Also, can you [rebase](https://www.simonemms.com/blog/2020/11/16/an-introduction-to-git-rebase) these changes as a single commit, please? I'd like to keep the commit messages in the same [format](https://www.conventionalcommits.org/en/v1.0.0/) - suggest `docs(readme): improve readme` as the commit message

**Comment by jgao1025** [comment]
> @jgao1025 sorry, I'd missed that you'd done updates. Thanks for the prompt. One last thing on there.
> 
> Also, can you [rebase](https://www.simonemms.com/blog/2020/11/16/an-introduction-to-git-rebase) these changes as a single commit, please? I'd like to keep the commit messages in the same [format](https://www.conventionalcommits.org/en/v1.0.0/) - suggest `docs(readme): improve readme` as the commit message

all done. :)

**Review by mrsimonemms — APPROVED** [review]
Thanks @jgao1025

**Pull request merged**